### DISCUSSION TITLE
Trying to implement timed license validation in app

### DISCUSSION THREAD

**Discussion by fmg-pete** [keygen-sh/community#52] [Q&A] [answered]
I am following [this guide,](https://keygen.sh/docs/choosing-a-licensing-model/timed-licenses/) but I seem to be missing something really fundamental in getting it working. There's no mention of machine activation or fingerprinting in the request flow, but it appears essential for it to work? I understand the fundamentals of fingerprinting, I just can't seem to figure out where and how it's implemented in the request flow. Do I need a third-party library fingerprint generating library?  Does this need to be applied to a newly created machine and then added to the license via another API call? When I use the C# code for the [Validate by license Key example](https://keygen.sh/docs/api/licenses/#licenses-actions-validate-key), I get Valid = False for the license, and "fingerprint is not activated (has no associated machine) ValidationCode=NO_MACHINE". Are there policy settings that negate the need for fingerprinting?

All the business wants is to be able to send a licensed user a key, have them copy / paste it in to the app somewhere, and the software to be unlocked once the key is validated ( offline or online ), and then leave them alone for a year.  Is there a simple way to do this? Thanks in advance...

**Comment by ezekg** [comment, marked as answer]
A timed license doesn't require machine activation, so it's not mentioned on that page. You can combine a timed license with a node-locked or floating license, which do require machine activation. But that doesn't sound like what you're looking for, so feel free to forego the activation step since it's not required for basic timed licenses, given you don't care about the number of machines each license can be used on. You can simply [validate the license key](https://keygen.sh/docs/api/licenses/#licenses-actions-validate-key) and be done with it.

You're receiving a `NO_MACHINE` validation code because you're sending a `scope.fingerprint` param in a license validation request, but the license has no machines activated yet. It sounds like you don't want to use activation, so you can forego that param and the validation should work, given the policy is not strict.

As far as fingerprinting — 

There are packages that can query the machine GUID across various platforms:

- For C# apps, you can access the machine GUID using the [DeviceId](https://github.com/MatthewKing/DeviceId) package via AddMachineGuid().
- For Python, we maintain a package called [py-machineid](https://github.com/keygen-sh/py-machineid).
- For Node, there's the [node-machine-id](https://github.com/automation-stack/node-machine-id) package.
- For Go, there's the [machineid](https://github.com/denisbrodbeck/machineid) package.

(More on activation and fingerprinting is covered [here](https://keygen.sh/docs/activating-machines/).)

Hope that helps? Let me know.

**Reply by fmg-pete** [reply]
Thanks very much for the info @ezekg, that was the answer.  I just had to remove the scope{...} from my JSON payload to the licenses/actions/validate-key endpoint. That all works now, latest struggle is getting the expiry timestamp which is not as straightforward as I'd hoped! I'll leave that fight for another day. :)

**Reply by ezekg** [reply]
No problem. The license's expiry is available in the license validation payload under `data.attributes.expiry`.

It'll be using the ISO8601 time format.

**Reply by fmg-pete** [reply]
I found it in the documentation, but I can't make sense of how to extract it from the payload. It's trivial in Javascript ( Response.Data.Attributes.Expiry ) but painful in C# because of the strict way it maps  JSON to variables.  Sure I'm missing something dumb, but this doesn't work: :(

```
public class Document<T>
{
     public T Data { get; set; }           
     public List<Error> Errors { get; set; } = new List<Error>();
     public Attributes Attributes { get; set; } = new Attributes();
}

public class Document<T, U> : Document<T>
{
     public U Meta { get; set; }
}

public class Attributes
{
     public string Expiry { get; set; }
}

```

**Reply by ezekg** [reply]
Try using something like https://json2csharp.com. It produced this output:

```csharp
// Root myDeserializedClass = JsonConvert.DeserializeObject<Root>(myJsonResponse);
public class Account
{
  public Links links { get; set; }
  public Data data { get; set; }
}

public class Attributes
{
  public object name { get; set; }
  public string key { get; set; }
  public DateTime expiry { get; set; }
  public string status { get; set; }
  public int uses { get; set; }
  public bool @protected { get; set; }
  public bool suspended { get; set; }
  public object scheme { get; set; }
  public bool encrypted { get; set; }
  public bool floating { get; set; }
  public bool strict { get; set; }
  public int maxMachines { get; set; }
  public object maxProcesses { get; set; }
  public int maxCores { get; set; }
  public object maxUses { get; set; }
  public bool requireHeartbeat { get; set; }
  public bool requireCheckIn { get; set; }
  public DateTime lastValidated { get; set; }
  public object lastCheckIn { get; set; }
  public object nextCheckIn { get; set; }
  public Metadata metadata { get; set; }
  public DateTime created { get; set; }
  public DateTime updated { get; set; }
}

public class Data
{
  public string id { get; set; }
  public string type { get; set; }
  public Links links { get; set; }
  public Attributes attributes { get; set; }
  public Relationships relationships { get; set; }
}

public class Entitlements
{
  public Links links { get; set; }
}

public class Error
{
  public string title { get; set; }
  public string detail { get; set; }
  public string code { get; set; }
  public Source source { get; set; }
}

public class Links
{
  public string self { get; set; }
  public string related { get; set; }
}

public class Machines
{
  public Links links { get; set; }
  public Meta meta { get; set; }
}

public class Meta
{
  public DateTime ts { get; set; }
  public bool valid { get; set; }
  public string detail { get; set; }
  public string code { get; set; }
  public int nonce { get; set; }
  public Scope scope { get; set; }
  public int count { get; set; }
}

public class Metadata
{
}

public class Policy
{
  public Links links { get; set; }
  public Data data { get; set; }
}

public class Product
{
  public Links links { get; set; }
  public Data data { get; set; }
}

public class Relationships
{
  public Account account { get; set; }
  public Product product { get; set; }
  public Policy policy { get; set; }
  public User user { get; set; }
  public Machines machines { get; set; }
  public Tokens tokens { get; set; }
  public Entitlements entitlements { get; set; }
}

public class Root
{
  public Meta meta { get; set; }
  public Data data { get; set; }
  public List<Error> errors { get; set; }
}

public class Scope
{
  public string fingerprint { get; set; }
}

public class Source
{
  public string pointer { get; set; }
}

public class Tokens
{
  public Links links { get; set; }
}

public class User
{
  public Links links { get; set; }
  public Data data { get; set; }
}
```

Then you can access `myDeserializedClass.Data.Attributes.Expiry`.

**Reply by fmg-pete** [reply]
Thanks again @ezekg for the detailed reply. I was resisting generating the entire API response class definition just to get the valid flag and the expiry date, but I tried it today... with the same result.  It appears none of the Attributes ( expiry, status, lastValidated, etc. ) get populated. There are no errors, but I suspect c# is screwing up the mapping from the response to the class.  I've seen this before where json2csharp's guess at the object type isn't quite right. Also breaks all the class mappings in [this example](https://github.com/keygen-sh/example-csharp-license-activation/blob/master/Program.cs). Unfortunately, Visual Studio's debugging for async requests is not much help either. 

Do you know of any complete c# examples that return those attribute values via an async request?  My starting point was just adding "expiry", but that's really led me down a bit of a rabbit hole!

Edit:

I can't even get this to work now, locally or otherwise.

https://dotnetfiddle.net/ZNLYqX ( account ID and keys obfuscated )

It's from this example ( Validate License by key ) :

https://keygen.sh/docs/api/licenses/

Have I exceeded the API request quota, or triggered  some sort of lock-out?  I just can't seem to get anything sensible from the API today, I'm sure it worked last week :(.

**Reply by ezekg** [reply]
I updated [this C# example](https://github.com/keygen-sh/example-csharp-license-validation/blob/master/Program.cs) to use `records` for deserialization. Hope that helps!

**Reply by fmg-pete** [reply]
Thanks for that @ezekg. I actually got a version working this morning using NewtonsoftJson deserialization and a dynamic object type. This lets you get at arbitrary nested keys in a JSON response object without having to build a model.  I can now get meta.valid, data.attribute.expiry and  data.attribute.key values ( or any value ) and use them for validation. Works fine in debug but the RestSharp RestRequest is not playing nicely in the release. I suspect it's blocking and not actually asynchronous, or throwing some exception I'm not catching... but that's down to my implementation - nothing to do with the API!

Once I (finally) have a proper async / await validate() function working I'll share it here.

**Reply by ezekg** [reply]
Good to hear, and good luck. Just FYI, I also updated the example above to use async/await.

**Reply by fmg-pete** [reply]
Finally, working. There were A LOT of gotchas, final hurdle was that it worked in Debug mode but not in Release. Turns out there are issues with the latest version of RestSharp. I eventually got it working on v105.15.0, API calls just never completed successfully on 107.x.x or 108.x.x unless I ran them in Visual Studio 2022 in Debug mode. 

Here's my function in c#:

        ```
       using RestSharp;
       using Newtonsoft.Json;
 

       async public static Task<IRestResponse> validate(string accountID, string licenseKey)
        {           

            var client = new RestClient("https://api.keygen.sh/v1/accounts/" + accountID);
            var request = new RestRequest("licenses/actions/validate-key").AddJsonBody(
            new
            {
                meta = new
                {
                    key = licenseKey
                }
            });
          
            var p = await client.ExecutePostAsync(request, default);
            string JsonResponse = p.Content.ToString();
            dynamic Validation = JsonConvert.DeserializeObject(JsonResponse);
            if (Validation.meta.valid == false)
            {              
               // Prompt user that licence is not valid
            } else {
               // do whatever with any of the values from the license validate by key API call
               // https://keygen.sh/docs/api/licenses/#licenses-actions-validate-key 
            }

            return p;
        }
```
